src/Cast.hpp

```diff
@@ -0,0 +1,199 @@
+#pragma once
+
+#include <Types.h>
+#include <BwString.h>
+#include <BWAPI/Position.h>
+#include <BWAPI/UnitType.h>
+#include <BWAPI/Event.h>
+#include <BWAPI/EventType.h>
+
+//____________________________________________//
+// Forward conversion :: BW -> RS
+// Reverse conversion :: RS -> BW
+// BW - types that come from BWAPI
+// RS - types that come from outside, e.g. Rust
+// for value types it looks like this: Cast<BWAPI::Position, Position>
+// for opaque ref types looks like this: Cast<BWAPI::Unit, Unit*>
+
+template<class BW, class RS>
+struct Cast {
+    // type members to infer Cast type instance from CastFwd and CastRev
+    typedef RS RsType;
+    typedef BW BwType;
+    // disallow to call non-specialized version
+    static RS from_bw(const BW& bw) = delete;
+    static BW to_bw(const RS& rs) = delete;
```

Comment:
In this case I may suggest using `T` instead of `RS` with the meaning some type as well as template parameter.